### DISCUSSION TITLE
Deduplication technique could be used in design

### DISCUSSION THREAD

**Discussion by yuanjingsong** [juicedata/juicefs#2129] [Ideas]
**What would you like to be added**:
Data deduplication is a technique that reuses the repeated data chunk (block in this project) among different files, and it has been used in Ceph project. 
There are several file systems use the deduplication to improve read performance and save storage cost.  [ATC 19'](https://www.usenix.org/conference/atc19/presentation/duggal), [TOS 19'](https://dl.acm.org/doi/abs/10.1145/3295461) 

**Why is this needed**:
1. The design of splitting files into chunks and managing metadata by database is similar to the deduplication design.
2. Deduplication could save disk cost and improve read performance.

**Comment by davies** [comment]
Data dedup is usually expensive to use, have few value to common use cases. 

For some cases that have lots of duplicated data, may be we can use hard link, copy_file_range(), snapshot (planned) to avoid the duplicated data, also much faster.

**Reply by yuanjingsong** [reply]
Okay, thanks for reply.

**Reply by vipsland** [reply]
i think u need to think blake3 instead of sha256. blake3 will be very cheap

**Reply by flibustenet** [reply]
Dedup can reduce network and number of blocks (on cache and on storage). 
When juicefs already use block and index where do you see an expensive downside ?

**Comment by vipsland** [comment]
Feature request : pls have a plugin whereby we can write our own interfacing for dedup etc, change our own hash to blake3 etc

> Data dedup is usually expensive to use... i dont agree.

it's expensive not to deploy

> For some cases that have lots of duplicated data, may be we can use hard link, copy_file_range(), snapshot (planned) to avoid the duplicated data, also much faster.

actually if zfs is compatible will have built in dedup without this additional part BUT still we shld code our own dedup too.
